src/main/kotlin/calculator/Calculator.kt

```diff
@@ -0,0 +1,71 @@
+package calculator
+
+fun String?.tokenize(): List<String> = this?.trim()?.split("\\s+".toRegex())?.filter { it != "" }
+    ?: throw IllegalArgumentException("Null cannot be tokenized")
+
+fun List<String>.parse(): Expression {
+    fun parse(exp: Expression, tokenList: List<String>): Expression {
+        if (tokenList.isEmpty()) return exp
+
+        val newExp = when (val parsed = Expression.parse(tokenList.first())) {
+            is Expression.Operand -> when (exp) {
+                Expression.Undefined -> parsed
+                is Expression.Operator -> exp.also { it.secondOperand = parsed }
+                else -> throw IllegalArgumentException("Invalid expression")
+            }
+            is Expression.Operator -> parsed.also { it.firstOperand = exp }
+            else -> throw IllegalArgumentException("Invalid expression")
+        }
+
+        return parse(newExp, tokenList.drop(1))
+    }
+
+    return parse(Expression.Undefined, this)
+}
+
+fun String?.calculate(): Int = this.tokenize().parse().eval()
+
+sealed class Expression {
+
+    abstract fun eval(): Int
+
+    object Undefined : Expression() {
+
+        override fun eval(): Int {
+            throw IllegalArgumentException("Incomplete expression")
+        }
+    }
+
+    data class Operator(
+        val function: (Int, Int) -> Int,
+        var firstOperand: Expression = Undefined,
+        var secondOperand: Expression = Undefined
+    ) : Expression() {
+
+        override fun eval(): Int {
+            return function(firstOperand.eval(), secondOperand.eval())
+        }
+    }
+
+    data class Operand(private val value: Int) : Expression() {
+        override fun eval(): Int = value
+    }
+
+
+    companion object {
+
+        fun parse(input: String): Expression {
+            return when (input) {
+                "+" -> Operator(Int::plus)
+                "-" -> Operator(Int::minus)
+                "*" -> Operator(Int::times)
+                "/" -> Operator(Int::div)
+                else -> try {
+                    Operand(input.toInt())
+                } catch (e: NumberFormatException) {
+                    throw IllegalArgumentException(e)
+                }
```

Comment:
```suggestion
                else -> {
                    val parsed = input.toIntOrNull() ?: throw IllegalArgumentException()
                    Operand(parsed)
                }
```
이 부분은 취향 차겠지만 앞으로 `xxxOrNull()` + 엘비스 연산자(?:)를 사용한 예외처리를 많이 보시게 될 것 같아서 남기는 코멘트입니다. 코틀린의 많은 API들이 `xxxOrNull`과 같은 네이밍으로 예외가 발생했을 때 null을 리턴하는데요, 커스텀 예외를 발생시키고 싶다면 자주 사용하게 되는 것 같아요 😄

Reply:
try 보다 엘비스가 깔끔하네요.
input.toIntOrNull()?.let { Operand(it) } ?: throw IllegalArgumentException()
로 parsed 로컬 변수까지 없애는건 너무 나간걸까요? ;)

Reply:
적절하게 `let`을 사용한 예시인 것 같아요 👍 

체이닝으로 표현하는 것과 적절하게 변수로 끊어주며 코딩하는 방식은 취향에 따라 적절하게 구현해주시면 될 것 같습니다!
저 같은 경우는 예전에는 체이닝을 선호하다가 최근에는 적당히 끊어가며 작성하는 편이에요 😄